ThreadPool.hpp

```diff
@@ -0,0 +1,102 @@
+// Purpose: Thread pool
+
+// Based on https://github.com/progschj/ThreadPool
+
+#pragma once
+
+#include <vector>
+#include <queue>
+#include <memory>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <future>
+#include <functional>
+#include <stdexcept>
+
+class ThreadPool final
+{
+   public:
+      // Launches specified number of worker threads
+      ThreadPool (size_t threads = 1);
+      ~ThreadPool ();
+
+      // Not copyable
+      ThreadPool (const ThreadPool &) = delete;
+      ThreadPool& operator= (const ThreadPool &) = delete;
+
+      // Not moveable
+      ThreadPool (ThreadPool &&) = delete;
+      ThreadPool& operator= (const ThreadPool &&) = delete;
+
+      // Enqueue task and return std::future<>
+      template<typename Callable, typename... Args>
+      auto enqueue (Callable&& callable, Args&&... args) 
+         -> std::future<typename std::result_of<Callable (Args...)>::type>;
+
+      // Enqueue task without requiring capture of std::future<>
+      // Note: Best not to let exceptions escape the callable
+      template<typename Callable, typename... Args>
+      auto enqueueAndDetach (Callable&& callable, Args&&... args) 
+         -> void;
+
+   private:
+      // Keep track of threads, so they can be joined
+      std::vector<std::thread> workers;
+      // Task queue
+      std::queue<std::function<void ()>> tasks;
+ 
+      // Synchronization
+      using Lock = std::unique_lock<std::mutex>;
+      std::mutex queue_mutex;
+      std::condition_variable condition;
+      bool stop = false;
+};
+
+// Add a new work item to the pool
+template<typename Callable, typename... Args>
+auto ThreadPool::enqueue (Callable&& callable, Args&&... args) 
+   -> std::future<typename std::result_of<Callable (Args...)>::type>
+{
+   using return_t = typename std::result_of<Callable (Args...)>::type;
+   using task_t = std::packaged_task<return_t ()>;
+
+   auto task = std::make_shared<task_t> (std::bind (std::forward<Callable> (callable), std::forward<Args> (args)...));
+   std::future<return_t> result = task->get_future();
+
+   { // Critical section
+      Lock lock (queue_mutex);
+
+      // Don't allow an enqueue after stopping
+      if (stop)
+         throw std::runtime_error ("enqueue on stopped ThreadPool");
+
+      // Push work back on the queue
+      tasks.emplace ([task](){ (*task)(); });
+   } // End critical section
+
+   // Notify a thread that there is new work to perform
+   condition.notify_one ();
+   return result;
+}
+
+// Add a new work item to the pool
+template<typename Callable, typename... Args>
+auto ThreadPool::enqueueAndDetach (Callable&& callable, Args&&... args) 
+   -> void
+{
+   { // Critical section
+      Lock lock (queue_mutex);
+
+      // Don't allow an enqueue after stopping
+      if (stop)
+         throw std::runtime_error ("enqueue on stopped ThreadPool");
```

Comment:
I’m not sure I understand the question.

 

This would be throwing on a caller’s thread when the internal ‘stop’ flag is true, and the caller was trying to add a task.  If the caller is calling member functions while the object is being destroyed… that would be another sort of issue.

 

From: peter karasev [mailto:notifications@github.com] 
Sent: Thursday, June 16, 2016 11:56 AM
To: progschj/ThreadPool <ThreadPool@noreply.github.com>
Cc: Steven LeMay <steven.lemay@sbcglobal.net>; Author <author@noreply.github.com>
Subject: Re: [progschj/ThreadPool] Add an alternate method to enqueueAndDetach a task without requiring the capture of std::future<> (#31)

 

In ThreadPool.hpp <https://github.com/progschj/ThreadPool/pull/31#discussion_r67404012> :

> +   // Notify a thread that there is new work to perform
> +   condition.notify_one ();
> +   return result;
> +}
> +
> +// Add a new work item to the pool
> +template<typename Callable, typename... Args>
> +auto ThreadPool::enqueueAndDetach (Callable&& callable, Args&&... args) 
> +   -> void
> +{
> +   { // Critical section
> +      Lock lock (queue_mutex);
> +
> +      // Don't allow an enqueue after stopping
> +      if (stop)
> +         throw std::runtime_error ("enqueue on stopped ThreadPool");

@progschj <https://github.com/progschj>  isn't this going to cause a throw from destructor potentially? that seems to the be only time when "stop" is called. IIRC you can never cleanly catch such an exception -- DTOR throw is undefined behavior. 

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/progschj/ThreadPool/pull/31/files/f761d527a83a37003ecef53c94718818e60d8601#r67404012> , or mute the thread <https://github.com/notifications/unsubscribe/AKJadATwrIy1WhowY4pCzd32mUBhU6JSks5qMZxagaJpZM4GfmQ6> .  <https://github.com/notifications/beacon/AKJadA15Aixeyyd70laKpBGr_6yKx0KRks5qMZxagaJpZM4GfmQ6.gif>